Packages/com.unity.inputsystem/InputSystem/Plugins/UI/MouseModel.cs

```diff
@@ -0,0 +1,317 @@
+        /// </summary>
+        public ButtonDeltaState lastFrameDelta { get; private set; }
+
+        /// <summary>
+        /// Internal bookkeeping data used by the uGUI system.
+        /// </summary>
+        public InternalData internalData { get; set; }
+
+        /// <summary>
+        /// Set's this object to it's default, unused state.
+        /// </summary>
+        public void Reset()
+        {
+            lastFrameDelta = ButtonDeltaState.NoChange;
+            m_IsDown = false;
+
+            internalData.Reset();
+        }
+
+        /// <summary>
+        /// Call this on each frame in order to reset properties that detect whether or not a certain condition was met this frame.
+        /// </summary>
+        public void OnFrameFinished()
+        {
+            lastFrameDelta = ButtonDeltaState.NoChange;
+        }
+
+        /// <summary>
+        /// Fills a <see cref="PointerEventData"/> with this mouse button's internally cached values.
+        /// </summary>
+        /// <param name="eventData">These objects are used to send data through the uGUI system.</param>
+        public void CopyTo(PointerEventData eventData)
+        {
+            InternalData bookkeeping = internalData;
+            eventData.dragging = bookkeeping.isDragging;
+            eventData.clickTime = bookkeeping.pressedTime;
+            eventData.pressPosition = bookkeeping.pressedPosition;
+            eventData.pointerPressRaycast = bookkeeping.pressedRaycast;
+            eventData.pointerPress = bookkeeping.pressedGameObject;
+            eventData.rawPointerPress = bookkeeping.pressedGameObjectRaw;
+            eventData.pointerDrag = bookkeeping.draggedGameObject;
+        }
+
+        /// <summary>
+        /// Fills this object with the values from a <see cref="PointerEventData"/>.
+        /// </summary>
+        /// <param name="eventData">These objects are used to send data through the uGUI system.</param>
+        public void CopyFrom(PointerEventData eventData)
+        {
+            InternalData bookkeeping = internalData;
+            bookkeeping.isDragging = eventData.dragging;
+            bookkeeping.pressedTime = eventData.clickTime;
+            bookkeeping.pressedPosition = eventData.pressPosition;
+            bookkeeping.pressedRaycast = eventData.pointerPressRaycast;
+            bookkeeping.pressedGameObject = eventData.pointerPress;
+            bookkeeping.pressedGameObjectRaw = eventData.rawPointerPress;
+            bookkeeping.draggedGameObject = eventData.pointerDrag;
+            internalData = bookkeeping;
+        }
+
+        private bool m_IsDown;
+    }
+
+    public struct MouseModel
+    {
+        public struct InternalData
+        {
+            /// <summary>
+            /// This tracks the current GUI targets being hovered over.  Syncs up to <see cref="PointerEventData.hovered"/>.
+            /// </summary>
+            public List<GameObject> hoverTargets { get; set; }
```

Comment:
No Lists please. Either InlinedArray or plain array.

(Doesn't seem to be used, though)

Reply:
It's used with in HandleEnterAndExit.  Will see what I can do with inlined array, the number of entries in the list is really hard to predict or cap.

Reply:
If InlinedArray isn't appropriate, going with just a plain array plus capacity is fine.

BTW note re mutability. Doesn't matter for this API if it goes internal but no external APIs should expose mutable data structures (think we'll have to fix that in NoiseFilter, too).